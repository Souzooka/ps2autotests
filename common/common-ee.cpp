#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <kernel.h>

#define TEST_THREAD_PRIORITY   0x40
#define TEST_THREAD_STACK_SIZE 0x8000

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

u32 g_testThreadDoneSema = 0;

extern int test_main(int argc, char *argv[]);

char testThreadStack[TEST_THREAD_STACK_SIZE] __attribute__ ((aligned(16)));
void testThreadProc() {
	//TODO: Print begin/end traces
	//TODO: Pass arguments from main thread to the test function
	test_main(0, NULL);
	SignalSema(g_testThreadDoneSema);
}

int main(int argc, char *argv[]) {
	//A thread is created to ensure the execution environment is the same
	//across all possible boot methods (direct hw boot, ps2link, emulator, etc.)
	
	ee_sema_t semaInfo;
	memset(&semaInfo, 0, sizeof(ee_sema_t));
	semaInfo.init_count = 0;
	semaInfo.max_count  = 1;
	semaInfo.option     = 0;
	g_testThreadDoneSema = CreateSema(&semaInfo);
	
	u32 testThreadId = 0;
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func               = (void*)&testThreadProc;
	threadParam.initial_priority   = TEST_THREAD_PRIORITY;
	threadParam.stack_size         = TEST_THREAD_STACK_SIZE;
	threadParam.stack              = testThreadStack;
	testThreadId = CreateThread(&threadParam);
	
	StartThread(testThreadId, NULL);
	WaitSema(g_testThreadDoneSema);
	
	return 0;
}
