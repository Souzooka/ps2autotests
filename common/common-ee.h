#ifndef PS2AUTOTESTS_COMMON_EE_H
#define PS2AUTOTESTS_COMMON_EE_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _EE
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

#undef main
#define main test_main

void schedf(const char *format, ...);
void flushschedf();

#endif
