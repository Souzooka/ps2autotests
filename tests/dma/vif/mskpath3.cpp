#include <common-ee.h>
#include <assert.h>
#include <kernel.h>
#include <malloc.h>
#include <string.h>
#include <ee_regs.h>
#include "emit_vifcode.h"
#include "vifregs.h"

#define EE_GIF_STAT_M3P (0x02)

void send_simple_fifo_vif(volatile u128 *fifo, void *vifcode, int size) {
	// Must be aligned to 16 bytes.
	assert((size & 0xf) == 0 && (((u32)vifcode) & 0xf) == 0);

	u128 *vif128 = (u128 *)vifcode;
	for (int i = 0; i < size / 16; ++i) {
		*fifo = vif128[i];
	}
}

int main(int argc, char *argv[]) {
	printf("-- TEST BEGIN\n");

	printf("mask value (initial): %08X\n", *R_EE_GIF_STAT & EE_GIF_STAT_M3P);
	
	{
		VIF::Packet vifcode(16);
		vifcode.MSKPATH3(true);

		send_simple_fifo_vif(VIF::VIF1_FIFO, vifcode.Raw(), 16);

		printf("mask value (after mskpath3(true))): %08X\n", *R_EE_GIF_STAT & EE_GIF_STAT_M3P);
	}

	{
		VIF::Packet vifcode(16);
		vifcode.MSKPATH3(false);

		send_simple_fifo_vif(VIF::VIF1_FIFO, vifcode.Raw(), 16);

		printf("mask value (after mskpath3(false))): %08X\n", *R_EE_GIF_STAT & EE_GIF_STAT_M3P);
	}

	printf("-- TEST END\n");
	return 0;
}
