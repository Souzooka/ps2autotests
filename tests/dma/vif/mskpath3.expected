-- TEST BEGIN
mask value (initial): 00000000
mask value (after mskpath3(true))): 00000002
mask value (after mskpath3(false))): 00000000
-- TEST END
